lib/time/time.go

```diff
@@ -0,0 +1,462 @@
+var Module = &starlarkstruct.Module{
+	Name: ModuleName,
+	Members: starlark.StringDict{
+		"duration":      starlark.NewBuiltin("duration", parseDuration),
+		"location":      starlark.NewBuiltin("location", parseLocation),
+		"now":           starlark.NewBuiltin("now", now),
+		"struct":        starlark.NewBuiltin("struct", starlarkstruct.Make),
+		"time":          starlark.NewBuiltin("time", parseTime),
+		"sleep":         starlark.NewBuiltin("sleep", sleep),
+		"fromtimestamp": starlark.NewBuiltin("fromtimestamp", fromUnixTime),
+
+		"zero": Time{},
+
+		"nanosecond":  durInt(time.Nanosecond),
+		"microsecond": durInt(time.Microsecond),
+		"millisecond": durInt(time.Millisecond),
+		"second":      durInt(time.Second),
+		"minute":      durInt(time.Minute),
+		"hour":        durInt(time.Hour),
+	},
+}
+
+func durInt(d time.Duration) starlark.Int {
+	return starlark.MakeInt64(int64(d))
+}
+
+// LoadModule loads the time module.
+// It is concurrency-safe and idempotent.
+func LoadModule() (starlark.StringDict, error) {
+	return starlark.StringDict{
+		ModuleName: Module,
+	}, nil
+}
+
+// NowFunc is a function that generates the current time. Intentionally exported
+// so that it can be overridden, for example by applications that require their
+// Starlark scripts to be fully deterministic.
+var NowFunc = time.Now
+
+func parseDuration(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var s starlark.String
+	if err := starlark.UnpackArgs("duration", args, kwargs, "s", &s); err != nil {
+		return nil, err
+	}
+
+	d, err := time.ParseDuration(string(s))
+	if err != nil {
+		return nil, err
+	}
+
+	return Duration(d), nil
+}
+
+func parseLocation(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var x starlark.String
+	if err := starlark.UnpackArgs("location", args, kwargs, "x", &x); err != nil {
+		return nil, err
+	}
+	loc, err := time.LoadLocation(string(x))
+	if err != nil {
+		return nil, err
+	}
+
+	return starlark.String(loc.String()), nil
+}
+
+func sleep(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var x starlark.Int
+	if err := starlark.UnpackArgs("sleep", args, kwargs, "x", &x); err != nil {
+		return starlark.None, err
+	}
+	dur, ok := x.Int64()
+	if !ok {
+		return starlark.None, fmt.Errorf("invalid sleep value")
+	}
+
+	time.Sleep(time.Duration(dur))
+	return starlark.None, nil
+}
+
+func parseTime(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var (
+		x, location starlark.String
+		format      = starlark.String(time.RFC3339)
+	)
+	if err := starlark.UnpackArgs("time", args, kwargs, "x", &x, "format?", &format, "location", &location); err != nil {
+		return nil, err
+	}
+
+	if location == "" {
+		t, err := time.Parse(format.String(), x.String())
```

Comment:
format.String() looks incorrect if format has type starlark.String as it returns a string literal, as if by repr(x).